Reimplement actor isolation checking for referencing a declaration.

DougGregor opened the pull request:
Start collapsing the several implementations of actor isolation checking
into a single place that determines what it means to reference a declaration
from a given context, potentially supplying an instance for an actor. This
is partly cleanup, and partly staging for the implementation of the
Sendable restrictions introduced in SE-0338. The result of this check
falls into one of three categories:

* Reference occurs within the same concurrency domain (actor/task)
* Reference leaves an actor context to a nonisolated context (SE-0338)
* Reference enters the context of the actor, which might require a
combination of implicit async, implicit throws, and a "distributed" check.

Throughout this change I've sought to maintain the existing semantics,
even where I believe they are incorrect. The changes to the test cases
are not semantic changes, but reflect the unification of some
diagnostic paths that changed the diagnostic text but not when or how
those diagnostics are produced. Additionally, SE-0338 has not yet been
implemented, although this refactoring makes it easier to implement
SE-0338.

Use this new actor isolation checking scheme to implement the most
common actor-isolation checks for accessing both member and
non-member declarations from expressions.

DougGregor committed:
Reimplement actor isolation checking for referencing a declaration.

Start collapsing the several implementations of actor isolation checking
into a single place that determines what it means to reference a declaration
from a given context, potentially supplying an instance for an actor. This
is partly cleanup, and partly staging for the implementation of the
Sendable restrictions introduced in SE-0338. The result of this check
falls into one of three categories:

* Reference occurs within the same concurrency domain (actor/task)
* Reference leaves an actor context to a nonisolated context (SE-0338)
* Reference enters the context of the actor, which might require a
combination of implicit async, implicit throws, and a "distributed" check.

Throughout this change I've sought to maintain the existing semantics,
even where I believe they are incorrect. The changes to the test cases
are not semantic changes, but reflect the unification of some
diagnostic paths that changed the diagnostic text but not when or how
those diagnostics are produced. Additionally, SE-0338 has not yet been
implemented, although this refactoring makes it easier to implement
SE-0338.

Use this new actor isolation checking scheme to implement the most
common actor-isolation check, which occurs when accessing a member of
an instance.

DougGregor commented:
@swift-ci please smoke test

ktoso commented on lib/AST/Decl.cpp:
```diff
@@ -9352,3 +9352,7 @@ void swift::simple_display(llvm::raw_ostream &out, AnyFunctionRef fn) {
+
+bool ActorIsolation::isDistributedActor() const {
+  return getKind() == ActorInstance && getActor()->isDistributedActor();
+}
```
❤️

ktoso commented on lib/Sema/TypeCheckConcurrency.h:
```diff
@@ -222,6 +222,151 @@ bool contextRequiresStrictConcurrencyChecking(
+      Optional<ReferencedActor> actorInstance = None);
+
+  operator Kind() const { return kind; }
+};
```
This entire way to model it is excellent, really looking forward to it 👍

DougGregor committed:
Switch non-member actor isolation checking to the new implementation

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Simplify 'inout' check for isolated actor state in async calls.

DougGregor committed:
Simplify keypath checking for actor isolation.

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor committed:
Simplify @objc checking for actor-isolated members.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Reseat the "is distributed thunk" check onto actor reference checking.

Rather than write an appoximation of the "is distributed thunk" check
within the type checker, use the new actor reference checking logic with
a referenced actor that is synthesized from the information in the
constraint system.

DougGregor committed:
Reimplement conformance isolation checking using the actor reference logic.

Reimplement the final client of ActorIsolationRestriction, conformance
isolation checking, to base it on the new "actor reference" logic.
Centralize the diagnostics emission so we have a single place where we
emit the primary diagnostic (which is heavily customized based on
actor isolation/distributed/etc.) and any relevant notes to make
adjustments to the witness and/or requirement, e.g., adding
'distributed', 'async', 'throws', etc. Improve the diagnostics
slightly by providing Fix-Its when suggesting that we add "async"
and/or "throws".

With the last client of ActorIsolationRestriction gone, remove it
entirely.

DougGregor commented:
@swift-ci please smoke test

ktoso commented on lib/Sema/CSApply.cpp:
```diff
@@ -7821,55 +7809,60 @@ bool ExprRewriter::isDistributedThunk(ConcreteDeclRef ref, Expr *context) {
+  case ActorReferenceResult::EntersActor:
+    return refResult.options.contains(ActorReferenceResult::Flags::Distributed);
   }
 }
```
This is awesome, thanks for centralizing all the logic 🥳
